flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/JobAutoScaler.java

```diff
@@ -17,15 +17,19 @@
 
 package org.apache.flink.kubernetes.operator.reconciler.deployment;
 
-import org.apache.flink.kubernetes.operator.api.AbstractFlinkResource;
 import org.apache.flink.kubernetes.operator.controller.FlinkResourceContext;
 
+import java.util.Map;
+
 /** Per-job Autoscaler instance. */
 public interface JobAutoScaler {
 
     /** Called as part of the reconciliation loop. Returns true if this call led to scaling. */
-    boolean scale(FlinkResourceContext<? extends AbstractFlinkResource<?, ?>> ctx);
+    boolean scale(FlinkResourceContext<?> ctx);
 
     /** Called when the custom resource is deleted. */
-    void cleanup(AbstractFlinkResource<?, ?> cr);
+    void cleanup(FlinkResourceContext<?> ctx);
+
+    /** Get the current parallelism overrides for the job. */
```

Comment:
Add the interface and `applyAutoscalerParallelismOverrides` can fix the this bug, however, it's a little complex and not clear.

As I understand, the applyAutoscalerParallelismOverrides is a part of scaling. Could we move it inside of the autoscaler?

If so, just keep the scale and cleanup is more clear here. And it can fix this bug directly if autoscaler support flink on yarn in the nearly future.

Reply:
We can keep your change if we're not sure whether it's general for all types of flink job.

We can discuss or refactor it if needed in the future due to it's not public Api.

Reply:
I think I prefer this being part of the autoscaling code but we can refactor at a later point in time.